enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,428 @@
+4. The machine API will render the nodes with the specified network device configuration.
+
+#### Scaling new Nodes with `machinesets`
+1. OpenShift administrator configures a machineset with `addressesFromPools` defined in the platform specification.
+2. OpenShift administrator or machine autoscaler scales machines
+3. If `addressesFromPools` contains a `AddressesFromPool` definition the machine controller will create an `IPAddressClaim`. The machine controller will:
+- Set a finalizer on the `IPAddressClaim` called `machines.openshift.io/ip-claim-protection`
+- Set an owner reference on the `IPAddressClaim` to the associated `Machine`.
+- Set the condition `IPAddressClaimed` of the associated `Machine` to indicate that it it awaiting IPAddressClaims to be bound
+- Block the creation of the underlying machine in the infrastructure until all associated `IPAddressClaim`s are bound
+4. An external controller will watch for `IPAddressClaim` instances that reference a address pool type known to the external controller.
+5. The external controller will create an `IPAddress` and bind it to its associated `IPAddressClaim`.  
+6. The machine controller will update the condition `IPAddressClaimed` of the associated `Machine` to indicate its `IPAddressClaim`(s) is bound.  If a `Machine` 
+has multiple associated `IPAddressClaims`, a single `IPAddressClaimed` condition will report the number of outstanding claims.
+7. The machine controller will then create the virtual machine with the network configuration in the network device spec and the `IPAddress`.
+
+~~~mermaid
+sequenceDiagram
+    machineset controller->>+machine: creates machine with<br> IPPool
+    machine controller-->machine controller: create IPAddressClaim<br>and wait for claim<br>to be bound
+    machine controller-->machine controller: IPAddressClaim ownerReference<br>refers to the machine
+    IP controller-->IP controller: processes claim and<br>allocates IP address    
+    IP controller-->IP controller: create IPAddress and bind IPAddressClaim
+    machine controller-->machine controller: build guestinfo.afterburn.initrd.network-kargs<br>and clone VM
+~~~
+
+On scale down:
+1. The machine controller will remove the finalizer on `IPAddressClaim` associated with a given `Machine` after the underlying virtual machine has been deleted.
+2. The kubernetes API will garbage collect the `IPAddressClaim` and `IPAddress` formerly associated with the `Machine`.
+3. Once the `IPAddress` associated with the machine is deleted, the external controller can reuse the address.
+
+In this workflow, the controller is responsible for managing, claiming, and releasing IP addresses.  
+
+A sample project [machine-ipam-controller](https://github.com/rvanderp3/machine-ipam-controller) is an example of a controller that implements this workflow.
+
+
+#### Variation [optional]
+
+### API Extensions
+
+2 additional CRDs will be introduced which to `machine.openshift.io`.
```

Comment:
I think we need to have a discussion about this vs just re-using the upstream API types, if we can use the upstream types it will be a lot less work and mean any IP implementations are compatible with CAPI too.

Some challenges to overcome:
* historically not allowed to use alpha APIs
* API migrations not really done in OpenShift, need to plan ahead for these
* Define ownership of the types, who/what is installing and lifecycling the upstream CAPI CRDs

CC @vincepri

Reply:
I think now we are just using the CAPI types directly, this section should be updated